### ISSUE TITLE
Only 64.07% Top1 accuracy for ResNet18 on 4w4a 

### ISSUE THREAD

**Issue by HaoKun-Li** [IST-DASLab/OBC#3]
Thanks for you wonderful work and available code!
However, when I use command `python main_trueobs.py rn18 imagenet quant --wbits 4 --abits 4 --save rn18_4w4a.pth ` to quantize the ResNet18 to 4w4a on Imagenet，I got a quantized model with 64.07% accuracy. This is much lower than the reported "69.56%" accuracy in Table 4 of the paper "Optimal Brain Compression: A Framework for Accurate Post-Training Quantization and Pruning". 
![image](https://github.com/IST-DASLab/OBC/assets/34161960/6f407352-8615-4c52-b58b-79ae4fcd7080)

Then, I try to use "Batchnorm tuning" and "Statistics correction" to recover the accuracy as the authors suggest.
With the command `python postproc.py rn18 imagenet rn18_4w4a.pth --bnt`,  the model gets "65.92%" accuracy.
With the command `python postproc.py rn18 imagenet rn18_4w4a.pth --statcorr --statcorr-samples 1024`,  the model gets "65.76%" accuracy.  
However, both of them still are much lower than the reported "69.56%" accuracy in the paper.

I also test the 8w8a quantized model and the performance is close to the original model as is shown in the following figure.
![image](https://github.com/IST-DASLab/OBC/assets/34161960/754d8c26-ba68-42fb-ae65-de68e48d967d)

Note that I use PyTorch v1.9.0 to conduct experiments. 

I want to know whether I run the wrong commands. 
Maybe I ignore some important details so that I get a much worse quantized model than that reported in the paper?

**Comment by efrantar** [comment]
Hi, in Table 4 we only perform weight but not activation quantization (for all methods); quantizing also all activations (including first and last layer) to 4-bit is a lot harder, so significantly lower accuracy is expected. Can you try only quantizing the weights and then running batchnorm tuning (which is our setup in this experiment as described in paragraph "Quantization") to see if this gives you a similar number to the one in Table 4?

**Comment by HaoKun-Li** [comment]
> Hi, in Table 4 we only perform weight but not activation quantization (for all methods); quantizing also all activations (including first and last layer) to 4-bit is a lot harder, so significantly lower accuracy is expected. Can you try only quantizing the weights and then running batchnorm tuning (which is our setup in this experiment as described in paragraph "Quantization") to see if this gives you a similar number to the one in Table 4?

Thanks!
When I run the command `python main_trueobs.py rn18 imagenet quant --wbits 4 --save rn18_4w.pth ` to quantize weight only, I get a model with 69.16% accuracy.
Then I use `python postproc.py rn18 imagenet rn18_4w.pth --bnt` to conduct the batchnorm tuning, and get a model with 69.45% accuracy, which is a similar number to the one in Table 4.
